### PR TITLE
feat(trpc-user): Create the tRPC backend and add user procedures

### PR DESCRIPTION
**Pull request by wsy19961129**
This PR:
1. Build the tRPC backend.
2. Add the tRPC backed for user-related APIs.

Note that as we agreed in https://github.com/ideafast/ideafast-portal/issues/2393, this PR focuses on adding backend functions and passing the build and unit tests.

Several related backend functions are also added to help pass the build and tests; such functions will be updated again later.

Please focus on the type definitions, core functions, tRPC procedures and the unit tests of 'user'.

### CODE REVIEW

File: packages/itmat-ui-react/webpack.config.js
Comment (fguitton) [inline packages/itmat-ui-react/webpack.config.js:1]: Why are changes on this files required ?

File: packages/itmat-types/src/types/file.ts [hunk 1]
```diff
@@ -1,14 +1,53 @@
...
     hash: string;
-    metadata: Record<string, unknown>
+    fileType: enumFileTypes;
+    fileCategory: enumFileCategories;
+    sharedUsers?: string[];
 }
+
+
+
+export enum enumFileTypes {
```
Comment (fguitton) [inline packages/itmat-types/src/types/file.ts:20]: What is the motivation behind maintaining a record for the file type ? I understand this only here to represent the extension of the file ? I'm not sure it is useful ?

**Review by fguitton — CHANGES_REQUESTED** [review]
I have two main point to raise on this PR.
1) I have big concerns about the success of data migration for older versions. I can see a large number of `m_*` variables have been renamed amongst other things, we need to be extra careful as to how this is done.
2) I am not fully understanding the concept of cores in this sense. You have essentially completely reimplemented everything twice. The architecture we discussed originally leveraged an architecture where the core package would provide the execution logic for both the TRPC and GraphQL layers. Taking as an example the `getUsers` function: They do completely different things, and each separately called the DB driver. That is really not a scalable way to address this issue. It is perhaps worth discussing this again ?

**Comment by wsy19961129** [comment]
> I have two main point to raise on this PR.
> 
> 1. I have big concerns about the success of data migration for older versions. I can see a large number of `m_*` variables have been renamed amongst other things, we need to be extra careful as to how this is done.
> 2. I am not fully understanding the concept of cores in this sense. You have essentially completely reimplemented everything twice. The architecture we discussed originally leveraged an architecture where the core package would provide the execution logic for both the TRPC and GraphQL layers. Taking as an example the `getUsers` function: They do completely different things, and each separately called the DB driver. That is really not a scalable way to address this issue. It is perhaps worth discussing this again ?

Hi Florian,
1. Yes, previously the `m_` variables include those that are only used in the IDEAFAST (such as subjected, and visited); I think for a more general situation we should not assume any pre-defined properties of the subject, thus I rename them. I think the necessary actions indeed are required, so it is risky but I guess I still prefer to make these changes.
2. I agree. In the roadmap,  what I plan to do is to keep both core functions at the early stage and merge them into one once both the two collections of APIs are working as expected (roadmap 2.17: merge tRPC and GraphQL core functions into one). What do you think of this?

File: packages/itmat-cores/src/GraphQLCore/studyCore.ts [hunk 2]
```diff
@@ -16,20 +16,20 @@ import { ObjectStore } from '@itmat-broker/itmat-commons';
...
+    metadata?: Record<string, unknown>
 }
 
 export interface EditFieldInput {
     fieldId: string;
     fieldName: string;
     tableName?: string;
-    dataType: enumValueType;
-    possibleValues?: IValueDescription[]
+    dataType: string;
```
[the remark this answers is not in the document]
Comment (wsy19961129) [reply packages/itmat-cores/src/GraphQLCore/studyCore.ts:31]: It is an error, will fix this.

File: packages/itmat-types/src/types/file.ts (hunk 1 above)
Comment (wsy19961129) [reply packages/itmat-types/src/types/file.ts:20]: For thie PR it is more like a placeholder. 
I use that for validating the file type if needed (e.g., v3 allows upload profile (images) so only JPG, PNG, etc could be uploaded); and in further if there is a need to parse the file then we can design parsing methods based on the file extension.

**Pull request merged**